spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -220,6 +232,9 @@ private void addChannelFromString(Collection<MessageChannel> channels, String ch
 		MessageChannel channel = resolveChannelForName(channelName, message);
 		if (channel != null) {
 			channels.add(channel);
+			if (!mapped && !this.dynamicChannels.contains(channelName)) {
+				this.dynamicChannels.add(channelName);
```

Comment:
If we route at least once, then I think it should remain. Maybe we should make them weak references in case there's a corner case where someone creates channels dynamically.